### PR TITLE
Single phase flow with a 2D network of fractures

### PR DESCRIPTION
**Pull request by Cirdans-Home**
## Proposed changes

Implementation of a Single-Phase flow test with a 2D fracture network and different material conditions/tangent/normal permeabilities.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [x] Other: added example

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

**Comment by keileg** [comment]
Dear Fabio,

Thanks a lot for taking the time to contribute the code, we really appreciate it! 

As previously mentioned, we would like to make some changes to the code before including it, such as making the geometry easier to access from outside this module, and doing some minor cleanup of the code. 

Is it okay with you if we include these changes as part of the PR? This should require no effort from your side, we will also take care of fixing questions that may arise during code review.

Best,
Eirik

**Comment by Cirdans-Home** [comment]
Dear Eirik,

Sure, go ahead. I tried to adapt it to the new version of the code, but obviously, it's still quite far from your style.

Fabio

File: src/porepy/applications/md_grids/fracture_sets.py [hunk 1]
```diff
@@ -65,13 +65,12 @@ def orthogonal_fractures_3d(size: pp.number) -> list[pp.PlaneFracture]:
     return [pp.PlaneFracture(pts[i]) for i in range(3)]
 
 
-def benchmark_regular_2d_fractures(
-    size: Optional[pp.number] = 1,
-) -> list[pp.LineFracture]:
-    """Return a list of regular fractures as used in the 2d benchmark study.
+def benchmark_regular_2d_fractures(size: pp.number = 1) -> list[pp.LineFracture]:
```
Comment (IvarStefansson) [inline src/porepy/applications/md_grids/fracture_sets.py:68]: @keileg, should we rename to benchmark_2d_case_3 for consistency with the new function below?

Comment (keileg) [reply src/porepy/applications/md_grids/fracture_sets.py:68]: Sounds like a good idea.

**Comment by keileg** [comment]
> Dear Eirik,
> 
> Sure, go ahead. I tried to adapt it to the new version of the code, but obviously, it's still quite far from your style.
> 
> Fabio

Thanks. We will pick it up from here then.

**Comment by IvarStefansson** [comment]
I have tried to incorporate the contribution into the benchmark tutorial. The geometry is accessible in `fracture_sets`. I also did some work on plot_grid and included the dimension-dependent permeability class in `constitutive_laws`.

**Review by keileg — COMMENTED** [review]
@IvarStefansson: Some comments here, others are comming from reviewnb. Also see my question in a separate channel regarding plot_grid.

File: src/porepy/viz/plot_grid.py [hunk 2]
```diff
@@ -122,33 +116,35 @@ def save_img(
 
 def plot_sd(
     sd: pp.Grid,
-    cell_value: Optional[np.ndarray],
-    vector_value: Optional[np.ndarray],
-    info: Optional[str],
+    cell_value: Optional[np.ndarray] = None,
+    vector_value: Optional[np.ndarray] = None,
+    info: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/viz/plot_grid.py:121]: I think (parts of) the file was not being checked at all. I don't know why, though.

**Comment by IvarStefansson** [comment]
<p>If you're doing something which is supported, it's really easy to use. Problem arises as soon as you want to do something unsupported - hacking can be tricky.</p>

---

 <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1081/discussion/#comment-1840919417'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by IvarStefansson** [comment]
<p>Thanks!</p>

---

 <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1081/discussion/#comment-1840919401'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: src/porepy/viz/plot_grid.py [hunk 3]
```diff
@@ -122,33 +116,35 @@ def save_img(
...
+        vector_value: vector values, one 3d vector for each cell or for each face (see
+            the _quiver function).
+        info: Which geometry information to display, see add_info.
+        kwargs: Keyword arguments:
             fig_size: Size of figure.
             fig_num: The number of the figure.
-            color_map: Limits of the cell value color axis.
+            fracturewidth_1d: Width of fractures in 1d.
+            color_map: Color map as str or matplotlib color map. Defaults to jet.
+            color_map_limits: Limits of the cell value color axis.
```
Comment (keileg) [inline src/porepy/viz/plot_grid.py:138]:
```suggestion
            color_map_limits: Limits of the cell value color axis, given as a numpy array
                with min and max values.
```suggestion
            color_map_limits: Limits of the cell value color axis.
```
```

**Comment by keileg** [comment]
<p>Then what.</p>

---

 <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1081/discussion/#comment-1840919367'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: src/porepy/models/constitutive_laws.py [hunk 4]
```diff
@@ -742,7 +742,93 @@ def normal_permeability(self, interfaces: list[pp.MortarGrid]) -> pp.ad.Operator
...
+        """Permeability [m^2].
+
+        This function combines the permeability of the matrix, fractures and
+        intersections.
+
+        Parameters:
+            subdomains: List of subdomains.
+
+        Returns:
+            Cell-wise permeability values.
```
Comment (OmarDuran) [inline src/porepy/models/constitutive_laws.py:769]:
```suggestion
            cell-wise permeability values.
```

File: src/porepy/models/constitutive_laws.py [hunk 5]
```diff
@@ -742,7 +742,93 @@ def normal_permeability(self, interfaces: list[pp.MortarGrid]) -> pp.ad.Operator
...
+            subdomains: List of subdomains.
+
+        Returns:
+            Cell-wise permeability operator.
+
+        """
+        return super().permeability(subdomains)
+
+    def fracture_permeability(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
+        """Permeability of fractures.
```
Comment (OmarDuran) [inline src/porepy/models/constitutive_laws.py:799]: Does this method encompass both in-plane and out-plane permeabilities? If so, please clarify it.

**Review by OmarDuran — APPROVED** [review]
A very valuable contribution. I have left a few minor comments.

File: src/porepy/models/constitutive_laws.py (hunk 5 above)
Comment (IvarStefansson) [reply src/porepy/models/constitutive_laws.py:799]: No, the out-of-plane permeability is handled in `normal_permeability`.

**Pull request merged**